retworkx/__init__.py

```diff
@@ -1730,3 +1733,65 @@ def _graph_union(
     merge_edges=False,
 ):
     return graph_union(first, second, merge_nodes=False, merge_edges=False)
+
+
+@functools.singledispatch
+def bfs_search(graph, source, visitor):
+    """Breadth-first traversal of a directed/undirected graph.
+
+    The pseudo-code for the BFS algorithm is listed below, with the annotated
+    event points, for which the given visitor object will be called with the
+    appropriate method.
+
+    ::
+
+        BFS(G, s)
+          for each vertex u in V
+              color[u] := WHITE
+          end for
+          color[s] := GRAY
+          EQUEUE(Q, s)                             discover vertex s
+          while (Q != Ø)
+              u := DEQUEUE(Q)
+              for each vertex v in Adj[u]          (u,v) is a tree edge
+                  if (color[v] = WHITE)
+                      color[v] = GRAY
+                  else                             (u,v) is a non - tree edge
+                      if (color[v] = GRAY)         (u,v) has a gray target
+                          ...
+                      else if (color[v] = BLACK)   (u,v) has a black target
+                          ...
+              end for
+              color[u] := BLACK                    finish vertex u
+          end while
```

Comment:
I think having this in the retworkx-core function's docstring would be nice too.

Reply:
Done in aa27296